packages/css/src/plugins/at-rules/parse-at-rule.ts

```diff
@@ -0,0 +1,111 @@
+// Matches min-width, min-height, max-width, max-height,
+// min-device-width, min-device-height, max-device-width, max-device-height,
+// width, height, device-width, and device-height.
+const property = /(?:(?<property>((?:min|max)-)?(?:device-)?(?:width|height))\s*)/;
+const colon = /(?<colon>:\s*)/;
+
+const length_ = /(?<length>-?\d*\.?\d+)(?<lengthUnit>ch|em|ex|px|rem)?\s*/;
+const length2 = /(?<length2>-?\d*\.?\d+)(?<lengthUnit2>ch|em|ex|px|rem)?\s*/;
+
+/**
+ * Extracts and parses breakpoint information from a media query. We define breakpoints as
+ * the `min-width`/`max-width`/`min-height`/`max-width`/`width`/`height` parts of a media query.
+ *
+ * There are four situations that this function can handle:
+ *
+ * ### Situation one
+ *
+ *     <property>: <length><lengthUnit>
+ *
+ *     e.g. max-width: 200px
+ *
+ * ### Situation two
+ *
+ *     <length><lengthUnit> <comparisonOperator> <width|height>
+ *
+ *     e.g. 200px >= width
+ *
+ * ### Situation three
+ *
+ *     <width|height> <comparisonOperator> <length><lengthUnit>
+ *
+ *     e.g. width <= 200px
+ *
+ * ### Situation four
+ *
+ *     <length><lengthUnit> <comparisonOperator> <width|height> <comparisonOperator2> <length2><lengthUnit2>
+ *
+ *     e.g. 0 <= width <= 200px
+ *
+ * Note that more exotic syntax (e.g. `calc()` functions, ratios) are not currently
+ * supported.
+ *
+ * They will either be returned without being parsed, or throw an error.
+ *
+ * @param params The original media query, as a string
+ * @returns The extracted and parsed breakpoints from the media query
+ */
+export const parseAtRule = (params: string): ParsedAtRule[] => {
+  // Inspired by previous work from
+  // https://github.com/OlehDutchenko/sort-css-media-queries/blob/master/lib/create-sort.js
+
+  const parsedMatches: ParsedAtRule[] = [];
+
+  const situations: Situations = [
+    // Situation one
+    {
+      regex: property.source + colon.source + length_.source,
+      parser: parseSituationOne,
+    },
+
+    // Situation two
+    {
+      regex: length_.source + comparisonOperators.source + property.source,
+      parser: parseSituationTwo,
+    },
+
+    // Situation three
+    {
+      regex: property.source + comparisonOperators.source + length_.source,
+      parser: parseSituationThree,
+    },
+
+    // Situation four
+    {
+      regex:
+        length_.source +
+        comparisonOperators.source +
+        property.source +
+        comparisonOperators2.source +
+        length2.source,
+      parser: parseSituationFour,
+    },
+  ];
+
+  for (const { regex, parser } of situations) {
+    const matches = [...params.matchAll(new RegExp(regex, 'g'))];
+    for (const match of matches) {
+      const parsedMatch = parser(match);
+      if (parsedMatch) {
+        parsedMatches.push(parsedMatch);
+      }
+    }
+  }
+
+  // Sort matches from first to last, as they appear in the at-rule / media query
+  parsedMatches.sort((a, b) => {
```

Comment:
What is this sort needed for? Vs. the sort we do in the plugin below.

Reply:
this isn't really a sorting step 😅 it just makes sure that if we get cases like

```css
@media (width > 400px) or (width < 200px) { /* ... */ }
```

we want `parsedMatches` to preserve the order of the `width` conditions... in other words, we want to return something like `['width > 400px', 'width < 200px']`, and not `['width < 200px', 'width > 400px']`

this is so that when we compare this against another media query, e.g.

```css
@media (max-width: 200px) and (min-width: 100px) { /* ... */ }
```

we'll compare `width > 400px` with `max-width: 200px`, and `width < 200px` with `min-width: 100px` (after the two are parsed)


i'll add this to the comment for clarity

Reply:
Could this have been a `.reverse()`?

Reply:
ah sorry, i left off a bit of important context in my previous message

the `for (const { regex, parser } of situations) { ... }` loop above this means that any parts of the media query that match situation one (e.g. `min-width: 300px`) will appear in `parsedMatches` before any features that match situation two (e.g. `200px >= height`), which will appear before any features that match situations three and four (e.g. `0 <= width <= 200px`)

so if we have `@media (0 <= width <= 200px) and (200px >= height) or (min-width: 300px)`, then the three matches we found in the media query will go out of order - so we need to re-sort them to restore the correct order